### PR TITLE
Refactor researcher context and CPID harvesting

### PR DESCRIPTION
**Pull request by cyrossignol**
Part 2 of the changes started with #1477. 

These changes move the logic behind `HarvestCPIDs()` into a new set of classes that manage the BOINC researcher context and scraps the `GetNextProject()` behavior. It continues the CPID-related clean-up that I started in the previous PR. I merged @denravonska's changes from #1352--it helped to see what I was doing more clearly.

I added a provision that checks `REQUIRE_TEAM_WHITELIST_MEMBERSHIP` to prepare for team requirement removal. This isn't dynamic right now--it requires a wallet restart or a call to `resetcpids`. If needed, we can determine how to make the wallet handle this automatically with the contract handler changes.

This also fixes the empty "status" field on the overview screen. I'm currently reloading testnet and mainnet to check compatibility.

### CODE REVIEW

File: src/neuralnet/cpid.h [hunk 1]
```diff
@@ -344,4 +344,36 @@ class MiningId
...
+struct hash<NN::Cpid>
+{
+    //!
+    //! \brief Create a hash of the supplied CPID object.
+    //!
+    //! \param cpid Contains the bytes to hash.
+    //!
+    //! \return A hash as the sum of the two halves of the bytes in the CPID.
+    //!
+    size_t operator()(const NN::Cpid& cpid) const
```
Comment (denravonska) [inline src/neuralnet/cpid.h:370]: `size_t` is platform dependent (32 or 64 bit). Will that matter here?

Edit: Just saw the doc. Seems like it doesn't matter :)

File: src/neuralnet/researcher.cpp [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+{
+    const boost::optional<std::string> client_state_xml = ReadClientStateXml();
+
+    if (!client_state_xml) {
+        return { };
+    }
+
+    std::vector<std::string> projects = split(*client_state_xml, "<project>");
+
+    if (projects.size() < 2) {
```
Comment (denravonska) [inline src/neuralnet/researcher.cpp:115]: What happens if you only have one project?

File: src/neuralnet/cpid.h (hunk 1 above)
Comment (cyrossignol) [reply src/neuralnet/cpid.h:370]: I don't think it matters how much it wraps...from what I understand, `unordered_map` does an equality check after the hash lookup.

File: src/neuralnet/researcher.cpp (hunk 2 above)
Comment (cyrossignol) [reply src/neuralnet/researcher.cpp:115]: The first element from the split is just a bunch of crap from the beginning of the file. It doesn't contain a project, so we drop it right below this.

Comment (cyrossignol) [reply src/neuralnet/researcher.cpp:115]: (If `split()` returns one item, it's the whole file and doesn't contain any projects.)

**Comment by cyrossignol** [comment]
Main and test nets finished syncing with these changes. I need to make a tweak, and then I'll move out of draft.

**Review by jamescowens — APPROVED** [review]

**Pull request merged**